README.md
# Thumbstick

> A virtual joystick for touch capable interfaces

[![npm](https://img.shields.io/npm/v/thumbstick.svg)](https://npmjs.org/package/thumbstick)
[![npm](https://img.shields.io/npm/dm/thumbstick.svg)](https://npmjs.org/package/thumbstick)

This project owes a great deal of debt to the rather unfortunately-named [NippleJS](https://github.com/yoannmoinet/nipplejs). For now this project is basically a straight fork, which I intend to keep up-to-date with upstream. That may eventually change.

# Table Of Contents

<details>

<!-- toc -->

-   [Install](#install)
-   [Demo](#demo)
-   [Usage](#usage)
-   [Options](#options)
    -   [`options.zone` defaults to 'body'](#optionszone-defaults-to-body)
    -   [`options.color` defaults to 'white'](#optionscolor-defaults-to-white)
    -   [`options.size` defaults to 100](#optionssize-defaults-to-100)
    -   [`options.threshold` defaults to 0.1](#optionsthreshold-defaults-to-01)
    -   [`options.fadeTime` defaults to 250](#optionsfadetime-defaults-to-250)
    -   [`options.multitouch` defaults to false](#optionsmultitouch-defaults-to-false)
    -   [`options.maxNumberOfThumbs` defaults to 1](#optionsmaxnumberofthumbs-defaults-to-1)
    -   [`options.dataOnly` defaults to false](#optionsdataonly-defaults-to-false)
    -   [`options.position` defaults to `{top: 0, left: 0}`](#optionsposition-defaults-to-top-0-left-0)
    -   [`options.mode` defaults to 'dynamic'.](#optionsmode-defaults-to-dynamic)
        -   [`'dynamic'`](#dynamic)
        -   [`'semi'`](#semi)
        -   [`'static'`](#static)
    -   [`options.restJoystick` defaults to true](#optionsrestjoystick-defaults-to-true)
    -   [`options.restOpacity` defaults to 0.5](#optionsrestopacity-defaults-to-05)
    -   [`options.catchDistance` defaults to 200](#optionscatchdistance-defaults-to-200)
-   [API](#api)
    -   [Thumbstick instance (manager)](#thumbstick-instance-manager)
        -   [`manager.on(type, handler)`](#managerontype-handler)
        -   [`manager.off([type, handler])`](#managerofftype-handler)
        -   [`manager.get(identifier)`](#managergetidentifier)
        -   [`manager.destroy()`](#managerdestroy)
        -   [`manager.ids`](#managerids)
        -   [`manager.id`](#managerid)
    -   [thumbstick instance (joystick)](#thumbstick-instance-joystick)
    -   [`joystick.on`, `joystick.off`](#joystickon-joystickoff)
    -   [`joystick.el`](#joystickel)
    -   [`joystick.show([cb])`](#joystickshowcb)
    -   [`joystick.hide([cb])`](#joystickhidecb)
    -   [`joystick.add()`](#joystickadd)
    -   [`joystick.remove()`](#joystickremove)
    -   [`joystick.destroy()`](#joystickdestroy)
    -   [`joystick.identifier`](#joystickidentifier)
    -   [`joystick.trigger(type [, data])`](#joysticktriggertype--data)
    -   [`joystick.position`](#joystickposition)
    -   [`joystick.frontPosition`](#joystickfrontposition)
    -   [`joystick.ui`](#joystickui)
-   [Events](#events)
    -   [manager only](#manager-only)
        -   [`added`](#added)
        -   [`removed`](#removed)
    -   [manager and joysticks](#manager-and-joysticks)
        -   [`start`](#start)
        -   [`end`](#end)
        -   [`move`](#move)
        -   [`dir`](#dir)
        -   [`plain`](#plain)
        -   [`shown`](#shown)
        -   [`hidden`](#hidden)
        -   [`destroyed`](#destroyed)
        -   [`pressure`](#pressure)
-   [Contributing](#contributing)
    -   [Commits](#commits)
    -   [Style](#style)
    -   [Workflow](#workflow)
    -   [Build](#build)

<!-- tocstop -->
</details>

## Install

```bash
npm install thumbstick --save

// OR

bower install thumbstick --save
```

---

## Demo

Check out the [demo here](http://lazerwalker.com/thumbstick/#demo).

---

## Usage

Import it the way you want into your project :

```javascript
// CommonJS
var manager = require("thumbstick").create(options);
```

```javascript
// AMD
define(["thumbstick"], function(thumbstick) {
    var manager = thumbstick.create(options);
});
```

```html
// Global
<script src="./dist/thumbstick.min.js"></script>
<script>
    var manager = thumbstick.create(options);
</script>
```

**:warning: NB :warning:** Your joystick's container **has** to have its CSS `position` property set, either `absolute`, `relative`, `static`, ....

---

## Options

You can configure your joystick in different ways :

```javascript
var options = {
    zone: Element, // active zone
    color: String,
    size: Integer,
    threshold: Float, // before triggering a directional event
    fadeTime: Integer, // transition time
    multitouch: Boolean,
    maxNumberOfThumbs: Number, // when multitouch, what is too many?
    dataOnly: Boolean, // no dom element whatsoever
    position: Object, // preset position for 'static' mode
    mode: String, // 'dynamic', 'static' or 'semi'
    restJoystick: Boolean,
    restOpacity: Number, // opacity when not 'dynamic' and rested
    catchDistance: Number // distance to recycle previous joystick in
    // 'semi' mode
};
```

All options are optional :sunglasses:.

### `options.zone` defaults to 'body'

The dom element in which all your joysticks will be injected.

```html
<div id="zone_joystick"></div>

<script type="text/javascript" src="./thumbstick.min.js"></script>
<script type="text/javascript">
    var options = {
        zone: document.getElementById('zone_joystick');
    };
    var manager = thumbstick.create(options);
</script>
```

This zone also serve as the mouse/touch events handler.

It represents the zone where all your joysticks will be active.

### `options.color` defaults to 'white'

The background color of your joystick's elements.

Can be any valid CSS color.

### `options.size` defaults to 100

The size in pixel of the outer circle.

The inner circle is 50% of this size.

### `options.threshold` defaults to 0.1

This is the strength needed to trigger a directional event.

Basically, the center is 0 and the outer is 1.

You need to at least go to 0.1 to trigger a directional event.

### `options.fadeTime` defaults to 250

The time it takes for joystick to fade-out and fade-in when activated or de-activated.

### `options.multitouch` defaults to false

Enable the multitouch capabilities.

If, for reasons, you need to have multiple thumbs into the same zone.

Otherwise it will only get one, and all new touches won't do a thing.

Please note that multitouch is off when in `static` or `semi` modes.

### `options.maxNumberOfThumbs` defaults to 1

If you need to, you can also control the maximum number of instance that could be created.

Obviously in a multitouch configuration.

### `options.dataOnly` defaults to false

The library won't draw anything in the DOM and will only trigger events with data.

### `options.position` defaults to `{top: 0, left: 0}`

An object that will determine the position of a `static` mode.

You can pass any of the four `top`, `right`, `bottom` and `left`.

They will be applied as any css property.

Ex :

-   `{top: '50px', left: '50px'}`
-   `{left: '10%', bottom: '10%'}`

### `options.mode` defaults to 'dynamic'.

Three modes are possible :

#### `'dynamic'`

-   a new joystick is created at each new touch.
-   the joystick gets destroyed when released.
-   **can** be multitouch.

#### `'semi'`

-   new joystick is created at each new touch farther than `options.catchDistance` of any previously created joystick.
-   the joystick is faded-out when released but not destroyed.
-   when touch is made **inside** the `options.catchDistance` a new direction is triggered immediately.
-   when touch is made **oustide** the `options.catchDistance` the previous joystick is destroyed and a new one is created.
-   **cannot** be multitouch.

#### `'static'`

-   a joystick is positionned immediately at `options.position`.
-   one joystick per zone.
-   each new touch triggers a new direction.
-   **cannot** be multitouch.

### `options.restJoystick` defaults to true

Reset the joystick's position when it enters the rest state.

### `options.restOpacity` defaults to 0.5

The opacity to apply when the joystick is in a rest position.

### `options.catchDistance` defaults to 200

This is only useful in the `semi` mode, and determine at which distance we recycle the previous joystick.

At 200 (px), if you press the zone into a rayon of 200px around the previously displayed joystick,
it will act as a `static` one.

### `options.lockX` defaults to false

Locks joystick's movement to the x (horizontal) axis

### `options.lockY` defaults to false

Locks joystick's movement to the y (vertical) axis

---

## API

### Thumbstick instance (manager)

Your manager has the following signature :

```javascript
{
    on: Function,                       // handle internal event
    off: Function,                      // un-handle internal event
    get: Function,                      // get a specific joystick
    destroy: Function,                  // destroy everything
    ids: Array                          // array of assigned ids
    id: Number                          // id of the manager
    options: {
        zone: Element,                  // reactive zone
        multitouch: Boolean,
        maxNumberOfThumbs: Number,
        mode: String,
        position: Object,
        catchDistance: Number,
        size: Number,
        threshold: Number,
        color: String,
        fadeTime: Number,
        dataOnly: Boolean,
        restJoystick: Boolean,
        restOpacity: Number
    }
}
```

#### `manager.on(type, handler)`

If you whish to listen to internal events like :

```javascript
manager.on("event#1 event#2", function(evt, data) {
    // Do something.
});
```

Note that you can listen to multiple events at once by separating
them either with a space or a comma (or both, I don't care).

#### `manager.off([type, handler])`

To remove an event handler :

```javascript
manager.off("event", handler);
```

If you call off without arguments, all handlers will be removed.

If you don't specify the handler but just a type, all handlers for that type will be removed.

#### `manager.get(identifier)`

An helper to get an instance via its identifier.

```javascript
// Will return the thumbstick instanciated by the touch identified by 0
manager.get(0);
```

#### `manager.destroy()`

Gently remove all thumbs from the DOM and unbind all events.

```javascript
manager.destroy();
```

#### `manager.ids`

The array of thumbs' ids under this manager.

#### `manager.id`

The incremented id of this manager.

### thumbstick instance (joystick)

Each joystick has the following signature :

```javascript
{
    on: Function,
    off: Function,
    el: Element,
    show: Function,         // fade-in
    hide: Function,         // fade-out
    add: Function,          // inject into dom
    remove: Function,       // remove from dom
    destroy: Function,
    identifier: Number,
    trigger: Function,
    position: {             // position of the center
        x: Number,
        y: Number
    },
    frontPosition: {        // position of the front part
        x: Number,
        y: Number
    },
    ui: {
        el: Element,
        front: Element,
        back: Element
    },
    options: {
        color: String,
        size: Number,
        threshold: Number,
        fadeTime: Number
    }
}
```

### `joystick.on`, `joystick.off`

The same as the manager.

### `joystick.el`

Dom element in which the joystick gets created.

```html
<div class="thumbstick">
    <div class="front"></div>
    <div class="back"></div>
</div>
```

### `joystick.show([cb])`

Will show the joystick at the last known place.

You can pass a callback that will be executed at the end of the fade-in animation.

### `joystick.hide([cb])`

Will fade-out the joystick.

You can pass a callback that will be executed at the end of the fade-out animation.

### `joystick.add()`

Add the joystick's element to the dom.

### `joystick.remove()`

Remove the joystick's element from the dom.

### `joystick.destroy()`

Gently remove this thumbstick from the DOM and unbind all related events.

### `joystick.identifier`

Returns the unique identifier of the joystick.

Tied to its touch's identifier.

### `joystick.trigger(type [, data])`

Trigger an internal event from the joystick.

The same as `on` you can trigger multiple events at the same time.

### `joystick.position`

The absolute position of the center of the joystick.

### `joystick.frontPosition`

The absolute position of the back part of the joystick's ui.

### `joystick.ui`

The object that store its ui elements

```html
{
    el: <div class="thumbstick"></div>
    back: <div class="back"></div>
    front: <div class="front"></div>
}
```

---

## Events

You can listen events both on the manager and all the joysticks.

But some of them are specific to its instance.

If you need to listen to each joystick, for example, you can :

```javascript
manager
    .on("added", function(evt, thumbstick) {
        thumbstick.on("start move end dir plain", function(evt) {
            // DO EVERYTHING
        });
    })
    .on("removed", function(evt, thumbstick) {
        thumbstick.off("start move end dir plain");
    });
```

### manager only

#### `added`

A joystick just got added.

Will pass the instance alongside the event.

#### `removed`

A joystick just got removed.

Fired at the end of the fade-out animation.

Will pass the instance alongside the event.

Won't be trigger in a `dataOnly` configuration.

### manager and joysticks

Other events are available on both the manager and joysticks.

When listening on the manager,
you can also target **a joystick in particular** by prefixing
the event with its identifier, **`0:start`** for example.

Else you'll get all events from all the joysticks.

#### `start`

A joystick is activated. (the user pressed on the active zone)

Will pass the instance alongside the event.

#### `end`

A joystick is de-activated. (the user released the active zone)

Will pass the instance alongside the event.

#### `move`

A joystick is moved.

Comes with data :

```javascript
{
    identifier: 0,              // the identifier of the touch/mouse that triggered it
    position: {                 // absolute position of the center in pixels
        x: 125,
        y: 95
    },
    force: 0.2,                 // strength in %
    distance: 25.4,             // distance from center in pixels
    pressure: 0.1,              // the pressure applied by the touch
    angle: {
        radian: 1.5707963268,   // angle in radian
        degree: 90
    },
    instance: Thumb            // the thumbstick instance that triggered the event
}
```

#### `dir`

When a direction is reached after the threshold.

Direction are split with a 45° angle.

```javascript
//     \  UP /
//      \   /
// LEFT       RIGHT
//      /   \
//     /DOWN \
```

You can also listen to specific direction like :

-   `dir:up`
-   `dir:down`
-   `dir:right`
-   `dir:left`

In this configuration only one direction is triggered at a time.

#### `plain`

When a plain direction is reached after the threshold.

Plain directions are split with a 90° angle.

```javascript
//       UP               |
//     ------        LEFT | RIGHT
//      DOWN              |
```

You can also listen to specific plain direction like :

-   `plain:up`
-   `plain:down`
-   `plain:right`
-   `plain:left`

In this configuration two directions can be triggered at a time,
because the user could be both `up` and `left` for example.

#### `shown`

Is triggered at the end of the fade-in animation.

Will pass the instance alongside the event.

Won't be trigger in a `dataOnly` configuration.

#### `hidden`

Is triggered at the end of the fade-out animation.

Will pass the instance alongside the event.

Won't be trigger in a `dataOnly` configuration.

#### `destroyed`

Is triggered at the end of destroy.

Will pass the instance alongside the event.

#### `pressure`

> MBP's [**Force Touch**](http://www.apple.com/macbook-pro/features-retina/#interact), iOS's [**3D Touch**](http://www.apple.com/iphone-6s/3d-touch/), Microsoft's [**pressure**](https://msdn.microsoft.com/en-us/library/hh772360%28v=vs.85%29.aspx) or MDN's [**force**](https://developer.mozilla.org/en-US/docs/Web/API/Touch/force)

Is triggered when the pressure on the joystick is changed.

The value, between 0 and 1, is sent back alongside the event.

---

## Contributing

Your help is more than welcome, I would be very honored to have you on my side.

Here are some very basic guidelines.

#### Commits

Please follow these [guidelines](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit) so your commits will be taken by the self-generated changelog.

#### Style

There are both [JSCS](http://jscs.info/) and [ESLint](http://eslint.org/) in the project.

To test your code against them simply run `npm run prebuild`.

We follow a **4 spaces** rule around here.

#### Workflow

You can use the available scripts if needed.

-   `npm run watch` will run the build each time a change is detected.
-   `npm run formatAndLint` will test the formatting and the linting of your code.
-   `npm start` will run a static server that will serve the [`test/`](./test) folder on [`localhost:8080`](http://localhost:8080).
-   `npm test` will test using CasperJS, you have to run `npm start` in another window to have a local server available to CasperJS.

#### Build

Once you're satisfied with your changes, you can also include a build.

1. `npm run build` to generate built files.
2. commit your build with the message `chore: new build`.
3. `npm version patch|minor|major` depending on your change. Changelog will be generated and bower's version synced and everything is automatically committed (not pushed though).


CHANGELOG.md
<a name="0.8.2"></a>

## [0.8.1](https://github.com/lazerwalker/thumbstick/compare/v0.8.1...v0.8.2) (2018-10-23)

-   Remove some unused dependencies that also contain security vulnerabilities.

<a name="0.8.1"></a>

## [0.8.1](https://github.com/lazerwalker/thumbstick/compare/v0.7.1...v0.8.1) (2018-10-23)

-   Rename the library to "thumbstick"

<a name="0.7.1"></a>

## [0.7.1](https://github.com/lazerwalker/thumbstick/compare/v0.7.0...v0.7.1) (2018-06-06)

<a name="0.7.0"></a>

# [0.7.0](https://github.com/lazerwalker/thumbstick/compare/v0.6.8...v0.7.0) (2018-06-06)

### Bug Fixes

-   register touchcancel and pointercancel events ([2c541a6](https://github.com/lazerwalker/thumbstick/commit/2c541a6))

### Features

-   **joystick:** add lockX and lockY options ([6baba97](https://github.com/lazerwalker/thumbstick/commit/6baba97))

<a name="0.6.8"></a>

## [0.6.8](https://github.com/lazerwalker/thumbstick/compare/v0.6.7...v0.6.8) (2017-12-19)

### Bug Fixes

-   add touchcancel handling as touchend for events coming from outside ([ebc501e](https://github.com/lazerwalker/thumbstick/commit/ebc501e)), closes [#61](https://github.com/yoannmoinet/nipplejs/issues/61) [#57](https://github.com/yoannmoinet/nipplejs/issues/57) [#33](https://github.com/yoannmoinet/nipplejs/issues/33) [#31](https://github.com/yoannmoinet/nipplejs/issues/31) [#30](https://github.com/yoannmoinet/nipplejs/issues/30)
-   update manager with changing identifier ([45ec1d7](https://github.com/lazerwalker/thumbstick/commit/45ec1d7)), closes [#64](https://github.com/yoannmoinet/nipplejs/issues/64)

### Features

-   add restJoystick option ([55b6482](https://github.com/lazerwalker/thumbstick/commit/55b6482))

<a name="0.6.7"></a>

## [0.6.7](https://github.com/lazerwalker/thumbstick/compare/v0.6.6...v0.6.7) (2016-10-27)

### Bug Fixes

-   update dependencies ([73bcda9](https://github.com/lazerwalker/thumbstick/commit/73bcda9))
-   **events:** trigger events with the correct id and not identifier ([17f0fe6](https://github.com/lazerwalker/thumbstick/commit/17f0fe6))
-   **identifier:** better identifier management ([ff44916](https://github.com/lazerwalker/thumbstick/commit/ff44916))

<a name="0.6.6"></a>

## [0.6.6](https://github.com/lazerwalker/thumbstick/compare/v0.6.5...v0.6.6) (2016-08-23)

### Bug Fixes

-   **bin:** adapt slashes to windows if needed ([7dffd7c](https://github.com/lazerwalker/thumbstick/commit/7dffd7c))
-   **bin:** add missing space ([48f6a1f](https://github.com/lazerwalker/thumbstick/commit/48f6a1f))
-   **bin:** make the copyToGhPages script windows compatible ([11da89a](https://github.com/lazerwalker/thumbstick/commit/11da89a))
-   **dataOnly:** conserve chainability even with dataOnly ([d1eddd0](https://github.com/lazerwalker/thumbstick/commit/d1eddd0))

<a name="0.6.5"></a>

## [0.6.5](https://github.com/lazerwalker/thumbstick/compare/v0.6.4...v0.6.5) (2016-08-15)

### Bug Fixes

-   **collection:** avoid any native touch action on the `options.zone` ([bea2c42](https://github.com/lazerwalker/thumbstick/commit/bea2c42))

<a name="0.6.4"></a>

## [0.6.4](https://github.com/lazerwalker/thumbstick/compare/v0.6.3...v0.6.4) (2016-07-20)

### Bug Fixes

-   **thumbstick:** remove size of container and apply on children only ([0033342](https://github.com/lazerwalker/thumbstick/commit/0033342)), closes [#32](https://github.com/yoannmoinet/nipplejs/issues/32)

<a name="0.6.3"></a>

## [0.6.3](https://github.com/lazerwalker/thumbstick/compare/v0.6.2...v0.6.3) (2016-05-29)

### Bug Fixes

-   **manager:** add missing closing bracket ([9b99637](https://github.com/lazerwalker/thumbstick/commit/9b99637))
-   **resize:** listen to resize via event instead of global handler ([14c28f1](https://github.com/lazerwalker/thumbstick/commit/14c28f1))

<a name="0.6.2"></a>

## [0.6.2](https://github.com/lazerwalker/thumbstick/compare/v0.6.1...v0.6.2) (2015-12-31)

<a name="0.6.1"></a>

## [0.6.1](https://github.com/lazerwalker/thumbstick/compare/v0.6.0...v0.6.1) (2015-12-26)

### Bug Fixes

-   **collection:** trigger `start` event before any potential direction ([c5441fd](https://github.com/lazerwalker/thumbstick/commit/c5441fd))
-   **thumbstick:** fragment control of previous directions ([790fa26](https://github.com/lazerwalker/thumbstick/commit/790fa26)), closes [#25](https://github.com/yoannmoinet/nipplejs/issues/25)

### Features

-   **thumbstick:** add `resetDirection()` to cancel previous directions ([66f6e3b](https://github.com/lazerwalker/thumbstick/commit/66f6e3b))

<a name="0.6.0"></a>

# [0.6.0](https://github.com/lazerwalker/thumbstick/compare/v0.5.6...v0.6.0) (2015-11-29)

### Bug Fixes

-   **thumbstick:** call off() after triggering ([bf8e5c3](https://github.com/lazerwalker/thumbstick/commit/bf8e5c3))
-   **thumbstick:** only trigger itself, not its manager ([6b39786](https://github.com/lazerwalker/thumbstick/commit/6b39786))
-   **super:** move internal declarations where needed ([a4c79c2](https://github.com/lazerwalker/thumbstick/commit/a4c79c2))
-   **utils:** return object on u.extend ([9197c00](https://github.com/lazerwalker/thumbstick/commit/9197c00))
-   clean identifier if not found in any collection ([3496395](https://github.com/lazerwalker/thumbstick/commit/3496395))
-   handle scrolling offset by resetting zone's box ([6ccd0b5](https://github.com/lazerwalker/thumbstick/commit/6ccd0b5))
-   remove useless injections ([a0b1d7e](https://github.com/lazerwalker/thumbstick/commit/a0b1d7e))
-   return the found thumbstick in `get()` ([439aa69](https://github.com/lazerwalker/thumbstick/commit/439aa69))
-   unbind document when collection is manually destroyed ([6e66b62](https://github.com/lazerwalker/thumbstick/commit/6e66b62))

### Features

-   **thumbstick:** add an incremental id to absolutely differentiate each ([598a5eb](https://github.com/lazerwalker/thumbstick/commit/598a5eb))

src/thumbstick.js
///////////////////////
///   THE THUMB    ///
///////////////////////

function Thumb (collection, options) {
    this.identifier = options.identifier;
    this.position = options.position;
    this.frontPosition = options.frontPosition;
    this.collection = collection;

    // Defaults
    this.defaults = {
        size: 100,
        threshold: 0.1,
        color: 'white',
        fadeTime: 250,
        dataOnly: false,
        restJoystick: true,
        restOpacity: 0.5,
        mode: 'dynamic',
        zone: document.body,
        lockX: false,
        lockY: false
    };

    this.config(options);

    // Overwrites
    if (this.options.mode === 'dynamic') {
        this.options.restOpacity = 0;
    }

    this.id = Thumb.id;
    Thumb.id += 1;
    this.buildEl()
        .stylize();

    // Thumb's API.
    this.instance = {
        el: this.ui.el,
        on: this.on.bind(this),
        off: this.off.bind(this),
        show: this.show.bind(this),
        hide: this.hide.bind(this),
        add: this.addToDom.bind(this),
        remove: this.removeFromDom.bind(this),
        destroy: this.destroy.bind(this),
        resetDirection: this.resetDirection.bind(this),
        computeDirection: this.computeDirection.bind(this),
        trigger: this.trigger.bind(this),
        position: this.position,
        frontPosition: this.frontPosition,
        ui: this.ui,
        identifier: this.identifier,
        id: this.id,
        options: this.options
    };

    return this.instance;
};

Thumb.prototype = new Super();
Thumb.constructor = Thumb;
Thumb.id = 0;

// Build the dom element of the Thumb instance.
Thumb.prototype.buildEl = function (options) {
    this.ui = {};

    if (this.options.dataOnly) {
        return this;
    }

    this.ui.el = document.createElement('div');
    this.ui.back = document.createElement('div');
    this.ui.front = document.createElement('div');

    this.ui.el.className = 'thumbstick collection_' + this.collection.id;
    this.ui.back.className = 'back';
    this.ui.front.className = 'front';

    this.ui.el.setAttribute('id', 'thumb_' + this.collection.id +
        '_' + this.id);

    this.ui.el.appendChild(this.ui.back);
    this.ui.el.appendChild(this.ui.front);

    return this;
};

// Apply CSS to the Thumb instance.
Thumb.prototype.stylize = function () {
    if (this.options.dataOnly) {
        return this;
    }
    var animTime = this.options.fadeTime + 'ms';
    var borderStyle = u.getVendorStyle('borderRadius', '50%');
    var transitStyle = u.getTransitionStyle('transition', 'opacity', animTime);
    var styles = {};
    styles.el = {
        position: 'absolute',
        opacity: this.options.restOpacity,
        display: 'block',
        'zIndex': 999
    };

    styles.back = {
        position: 'absolute',
        display: 'block',
        width: this.options.size + 'px',
        height: this.options.size + 'px',
        marginLeft: -this.options.size / 2 + 'px',
        marginTop: -this.options.size / 2 + 'px',
        background: this.options.color,
        'opacity': '.5'
    };

    styles.front = {
        width: this.options.size / 2 + 'px',
        height: this.options.size / 2 + 'px',
        position: 'absolute',
        display: 'block',
        marginLeft: -this.options.size / 4 + 'px',
        marginTop: -this.options.size / 4 + 'px',
        background: this.options.color,
        'opacity': '.5'
    };

    u.extend(styles.el, transitStyle);
    u.extend(styles.back, borderStyle);
    u.extend(styles.front, borderStyle);

    this.applyStyles(styles);

    return this;
};

Thumb.prototype.applyStyles = function (styles) {
    // Apply styles
    for (var i in this.ui) {
        if (this.ui.hasOwnProperty(i)) {
            for (var j in styles[i]) {
                this.ui[i].style[j] = styles[i][j];
            }
        }
    }

    return this;
};

// Inject the Thumb instance into DOM.
Thumb.prototype.addToDom = function () {
    // We're not adding it if we're dataOnly or already in dom.
    if (this.options.dataOnly || document.body.contains(this.ui.el)) {
        return this;
    }
    this.options.zone.appendChild(this.ui.el);
    return this;
};

// Remove the Thumb instance from DOM.
Thumb.prototype.removeFromDom = function () {
    if (this.options.dataOnly || !document.body.contains(this.ui.el)) {
        return this;
    }
    this.options.zone.removeChild(this.ui.el);
    return this;
};

// Entirely destroy this thumbstick
Thumb.prototype.destroy = function () {
    clearTimeout(this.removeTimeout);
    clearTimeout(this.showTimeout);
    clearTimeout(this.restTimeout);
    this.trigger('destroyed', this.instance);
    this.removeFromDom();
    this.off();
};

// Fade in the Thumb instance.
Thumb.prototype.show = function (cb) {
    var self = this;

    if (self.options.dataOnly) {
        return self;
    }

    clearTimeout(self.removeTimeout);
    clearTimeout(self.showTimeout);
    clearTimeout(self.restTimeout);

    self.addToDom();

    self.restCallback();

    setTimeout(function () {
        self.ui.el.style.opacity = 1;
    }, 0);

    self.showTimeout = setTimeout(function () {
        self.trigger('shown', self.instance);
        if (typeof cb === 'function') {
            cb.call(this);
        }
    }, self.options.fadeTime);

    return self;
};

// Fade out the Thumb instance.
Thumb.prototype.hide = function (cb) {
    var self = this;

    if (self.options.dataOnly) {
        return self;
    }

    self.ui.el.style.opacity = self.options.restOpacity;

    clearTimeout(self.removeTimeout);
    clearTimeout(self.showTimeout);
    clearTimeout(self.restTimeout);

    self.removeTimeout = setTimeout(
        function () {
            var display = self.options.mode === 'dynamic' ? 'none' : 'block';
            self.ui.el.style.display = display;
            if (typeof cb === 'function') {
                cb.call(self);
            }

            self.trigger('hidden', self.instance);
        },
        self.options.fadeTime
    );
    if (self.options.restJoystick) {
        self.restPosition();
    }

    return self;
};

Thumb.prototype.restPosition = function (cb) {
    var self = this;
    self.frontPosition = {
        x: 0,
        y: 0
    };
    var animTime = self.options.fadeTime + 'ms';

    var transitStyle = {};
    transitStyle.front = u.getTransitionStyle('transition',
        ['top', 'left'], animTime);

    var styles = {front: {}};
    styles.front = {
        left: self.frontPosition.x + 'px',
        top: self.frontPosition.y + 'px'
    };

    self.applyStyles(transitStyle);
    self.applyStyles(styles);

    self.restTimeout = setTimeout(
        function () {
            if (typeof cb === 'function') {
                cb.call(self);
            }
            self.restCallback();
        },
        self.options.fadeTime
    );
};

Thumb.prototype.restCallback = function () {
    var self = this;
    var transitStyle = {};
    transitStyle.front = u.getTransitionStyle('transition', 'none', '');
    self.applyStyles(transitStyle);
    self.trigger('rested', self.instance);
};

Thumb.prototype.resetDirection = function () {
    // Fully rebuild the object to let the iteration possible.
    this.direction = {
        x: false,
        y: false,
        angle: false
    };
};

Thumb.prototype.computeDirection = function (obj) {
    var rAngle = obj.angle.radian;
    var angle45 = Math.PI / 4;
    var angle90 = Math.PI / 2;
    var direction, directionX, directionY;

    // Angular direction
    //     \  UP /
    //      \   /
    // LEFT       RIGHT
    //      /   \
    //     /DOWN \
    //
    if (
        rAngle > angle45 &&
        rAngle < (angle45 * 3) &&
        !obj.lockX
    ) {
        direction = 'up';
    } else if (
        rAngle > -angle45 &&
        rAngle <= angle45 &&
        !obj.lockY
    ) {
        direction = 'left';
    } else if (
        rAngle > (-angle45 * 3) &&
        rAngle <= -angle45 &&
        !obj.lockX
    ) {
        direction = 'down';
    } else if (!obj.lockY) {
        direction = 'right';
    }

    // Plain direction
    //    UP                 |
    // _______               | RIGHT
    //                  LEFT |
    //   DOWN                |
    if (!obj.lockY) {
        if (rAngle > -angle90 && rAngle < angle90) {
            directionX = 'left';
        } else {
            directionX = 'right';
        }
    }

    if (!obj.lockX) {
        if (rAngle > 0) {
            directionY = 'up';
        } else {
            directionY = 'down';
        }
    }

    if (obj.force > this.options.threshold) {
        var oldDirection = {};
        for (var i in this.direction) {
            if (this.direction.hasOwnProperty(i)) {
                oldDirection[i] = this.direction[i];
            }
        }

        var same = {};

        this.direction = {
            x: directionX,
            y: directionY,
            angle: direction
        };

        obj.direction = this.direction;

        for (var i in oldDirection) {
            if (oldDirection[i] === this.direction[i]) {
                same[i] = true;
            }
        }

        // If all 3 directions are the same, we don't trigger anything.
        if (same.x && same.y && same.angle) {
            return obj;
        }

        if (!same.x || !same.y) {
            this.trigger('plain', obj);
        }

        if (!same.x) {
            this.trigger('plain:' + directionX, obj);
        }

        if (!same.y) {
            this.trigger('plain:' + directionY, obj);
        }

        if (!same.angle) {
            this.trigger('dir dir:' + direction, obj);
        }
    }
    return obj;
};
